src/shim.c

```diff
@@ -1090,6 +1105,55 @@ main(int argc, char **argv)
 		shim_debug("Could not register function for atexit");
 	}
 
+	/* create pipe to monitor parent from the child */
+	if (pipe(monitor_pipe) < 0) {
+		shim_error("Could not create shim monitor pipe: %s",
+			strerror (errno));
+		goto out;
+	}
+
+	ret = atexit(send_end_byte);
+	if (ret) {
+		shim_error("Could not register function send_end_byte()"
+			" for atexit");
+		goto out;
+	}
+
+	pid = fork ();
+	if (pid < 0) {
+		shim_error("Could not spawn the child: %s", strerror(errno));
+		goto out;
+	} else if (!pid) {
+		/* child */
+		close(monitor_pipe[1]);
+
+		/* block reading on monitor_pipe */
+		char buf;
+		ssize_t bytes = read(monitor_pipe[0], &buf, sizeof(end_byte));
+		if (bytes == 0) {
+			/* no bytes read means connection has been closed,
+			 * let's send SIGKILL to the process inside the VM
+			 */
+			shim_debug("Parent has terminated because of SIGKILL"
+				"/nForwarding the SIGKILL to the container"
+				" process");
+
+			signal_handler(SIGKILL);
+			handle_signals(&shim);
+
+			/* restore the terminal because atexit functions won't
+			 * be called
+			 */
```

Comment:
ok. thanks for clarifying. Would be useful to add the "with _exit()" to your above comment.
Other than that lgtm